test/Microsoft.Azure.SignalR.Management.Tests/E2ETest/ServiceHubContextE2EFacts.cs

```diff
@@ -18,63 +20,162 @@ public class ServiceHubContextE2EFacts
+            Func<Task> sendTaskFunc = () => serviceHubContext.Clients.Group(_groupNames[0]).SendAsync(MethodName, Message);
+            await RunTestCore(clientEndpoint, clientAccessTokens, () => SendToGroupCore(serviceHubContext, sendTaskFunc), _userNames.Length / _groupNames.Length + _userNames.Length % _groupNames.Length);
+        }
+
+        [ConditionalTheory]
+        [SkipIfConnectionStringNotPresent]
+        [InlineData(ServiceTransportType.Transient)]
+        internal async Task SendToGroupsTest(ServiceTransportType serviceTransportType)
+        {
+            var (clientEndpoint, clientAccessTokens, serviceHubContext) = await InitAsync(serviceTransportType);
+            Func<Task> sendTaskFunc = () => serviceHubContext.Clients.Groups(_groupNames).SendAsync(MethodName, Message);
+            await RunTestCore(clientEndpoint, clientAccessTokens, () => SendToGroupCore(serviceHubContext, sendTaskFunc), ClientConnectionCount);
+        }
+
+        private static async Task SendToGroupCore(IServiceHubContext serviceHubContext, Func<Task> sendTask)
+        {
+            await Task.WhenAll(from i in Enumerable.Range(0, _userNames.Length)
+                               select serviceHubContext.UserGroups.AddToGroupAsync(_userNames[i], _groupNames[i % _groupNames.Length]));
+            await Task.Delay(_timeout);
+            await sendTask();
+            await Task.Delay(_timeout);
+            await Task.WhenAll(from i in Enumerable.Range(0, _userNames.Length)
+                               select serviceHubContext.UserGroups.RemoveFromGroupAsync(_userNames[i], _groupNames[i % _groupNames.Length]));
+            await Task.Delay(_timeout);
+            await sendTask();
+            await Task.Delay(_timeout);
+        }
 
-            var connections = await CreateAndStartClientConnections(clientEndpoint, clientAccessToken);
-            var receivedMessageCount = 0;
-            ListenOnMessage(connections, () => receivedMessageCount++);
+        private static async Task RunTestCore(string clientEndpoint, IEnumerable<string> clientAccessTokens, Func<Task> taskFunc, int expectedReceivedMessageCount)
+        {
+            var connections = await CreateAndStartClientConnections(clientEndpoint, clientAccessTokens);
+            var receivedMessageCount = new StrongBox<int>();
+            var listenTasks = ListenOnMessage(connections, () => Interlocked.Increment(ref receivedMessageCount.Value));
 
             Task task = null;
             try
             {
-                await (task = serviceHubContext.Clients.All.SendAsync(MethodName, Message));
+                await (task = taskFunc.Invoke());
             }
             finally
             {
                 Assert.Null(task.Exception);
             }
 
-            await Task.Delay(_timeout);
+            await Task.WhenAny(Task.WhenAll(listenTasks), Task.Delay(_timeout));
```

Comment:
>Delay [](start = 63, length = 5)

reuse `OrTimeout` inside [https://github.com/Azure/azure-signalr/blob/dev/test/Microsoft.Azure.SignalR.Tests/Infrastructure/TaskExtensions.cs](https://github.com/Azure/azure-signalr/blob/dev/test/Microsoft.Azure.SignalR.Tests/Infrastructure/TaskExtensions.cs)

Reply:
Can't reuse the  `OrTimeout` here, since some test cases is that sender send to one of several users, while other users receive no messages. In this case, the timeout part will always be reached, then a timeout exception will be thrown, witch is not by design.